File: hayden/week7/2025-07-20-Permutations.py
# Question: Permutations (https://leetcode.com/problems/permutations/)

# Time Complexity: O(n * n!)
# Space Complexity: O(n!)

# Notes:
# - Use backtracking to generate all possible permutations.

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []

        def backtrack(path, remaining):
            if not remaining:
                result.append(path[:])
                return

            for i in range(len(remaining)):
                backtrack(path + [remaining[i]], remaining[:i] + remaining[i+1:])

        backtrack([], nums)
        return result

File: 2025/SoyoungJung/week7/2025-07-20-Permutations.ts
function permute(nums: number[]): number[][] {
  const result: number[][] = [];

  const backtrack = (path: number[], remaining: number[]) => {
    if (remaining.length === 0) {
      result.push(path);
      return;
    }

    for (let i = 0; i < remaining.length; i++) {
      const next = remaining.slice(0, i).concat(remaining.slice(i + 1));
      backtrack([...path, remaining[i]], next);
    }
  };

  backtrack([], nums);
  return result;
}
